omegaconf/built_in_resolvers.py

```diff
@@ -0,0 +1,167 @@
+import os
+import warnings
+
+# from collections.abc import Mapping, MutableMapping
+from typing import Any, List, Mapping, Optional, Union
+
+from ._utils import _DEFAULT_MARKER_, Marker, _get_value, decode_primitive
+from .base import Container
+from .basecontainer import BaseContainer
+from .dictconfig import DictConfig
+from .errors import ConfigKeyError, ValidationError
+from .grammar_parser import parse
+from .listconfig import ListConfig
+from .nodes import AnyNode
+from .omegaconf import OmegaConf
+
+# Special marker use as default value when calling `OmegaConf.select()`. It must be
+# different from `_DEFAULT_MARKER_`, which is used by `OmegaConf.select()`.
+_DEFAULT_SELECT_MARKER_: Any = Marker("_DEFAULT_SELECT_MARKER_")
+
+
+def decode(expr: Optional[str], _parent_: Container) -> Any:
+    """
+    Parse and evaluate `expr` according to the `singleElement` rule of the grammar.
+
+    If `expr` is `None`, then return `None`.
+    """
+    if expr is None:
+        return None
+
+    if not isinstance(expr, str):
+        raise TypeError(
+            f"`oc.decode` can only take strings or None as input, "
+            f"but `{expr}` is of type {type(expr).__name__}"
+        )
+
+    parse_tree = parse(expr, parser_rule="singleElement", lexer_mode="VALUE_MODE")
+    val = _parent_.resolve_parse_tree(parse_tree)
+    return _get_value(val)
+
+
+def dict_keys(
+    in_dict: Union[str, Mapping[Any, Any]],
+    _root_: BaseContainer,
+    _parent_: Container,
+) -> ListConfig:
+    in_dict = _get_and_validate_dict_input(
+        in_dict, root=_root_, resolver_name="oc.dict.keys"
+    )
+    assert isinstance(_parent_, BaseContainer)
+
+    ret = OmegaConf.create(list(in_dict.keys()), parent=_parent_)
+    assert isinstance(ret, ListConfig)
+    return ret
+
+
+def dict_values(
+    in_dict: Union[str, Mapping[Any, Any]], _root_: BaseContainer, _parent_: Container
+) -> ListConfig:
+    in_dict = _get_and_validate_dict_input(
+        in_dict, root=_root_, resolver_name="oc.dict.values"
+    )
+
+    if isinstance(in_dict, DictConfig):
+        # DictConfig objects are handled in a special way: the goal is to make the
+        # returned ListConfig point to the DictConfig nodes through interpolations.
+
+        dict_key: Optional[str] = None
+        if in_dict._get_root() is _root_:
+            # Try to obtain the full key through which we can access `in_dict`.
+            if in_dict is _root_:
+                dict_key = ""
+            else:
+                dict_key = in_dict._get_full_key(None)
+                if dict_key:
+                    dict_key += "."  # append dot for future concatenation
+                else:
+                    # This can happen e.g. if `in_dict` is a transient node.
+                    dict_key = None
```

Comment:
> ```python
> cfg = OmegaConf.create({})
> fake_child = OmegaConf.create({}, parent=cfg)
> ```
> 
> Even though fake_child is in the same root, an interpolation into fake_child is not possible.

But here `fake_child._get_full_key(None)` is equal to `""` so it's properly detected as invalid.
I guess someone could create a fake child with an invalid key, but at this point I think it's fair for some stuff to break.

`OmegaConf.select(_root_, dict_key) is not in_dict` wouldn't actually work for dynamically generated DictConfigs (in addition to add extra overhead). See the test `test_dict_values_dictconfig_resolver_output` for what I mean by that.

Reply:
> Select is just calling _get_node(key) iteratively with fragments of the key, it should be very efficient. (we only need to validate the input node, not all the generated interpolation nodes).

Just flagging that `select()` still has to resolve interpolations along the way (if any).

Maybe if we only support the "select" syntax, i.e. `${oc.dict.values:abc.def}`, then it will be easier since we directly know the correct path.

Reply:
Should be solved now with 27e59f8 (and companion commit 5a1f49c)